### DISCUSSION TITLE
Python Comprehensions in Cpp2

### DISCUSSION THREAD

**Discussion by msadeqhe** [hsutter/cppfront#757] [Suggestions]
Let's again glance terser function syntax:

```cpp
: (x) x + 1
```

Cpp2 already doesn't require parentheses within control structures, and it allows to write expressions without `=` in function declarations at the same time. These syntax choices made Cpp2's syntax somehow compatible with Python's _`if`-expressions_ and _Comprehensions_. What if Cpp2 also support them?

```cpp
x if x < 10 else x + 1
i for i in list if i < 10
```

For example:

```cpp
var1: i32 = 0 if is_off() else 1;

list: vector<i32> = (1, 2, 3, 4, 5);
var2: = i * 2 for i in list; // (2, 4, 6, 8, 10)
```

_Comprehensions_ are harder to implement than _`if`-expressions_. So it may be the reason to completely ignore them in Cpp2.

**Comment by AbhinavK00** [comment]
Comprehensions would eliminate the need for the ternary `?:` operator. And that would unify cpp's if statements and ternary expressions under one `if` in cpp2.

**Comment by hsutter** [comment]
Re ternary `?:` operator: Yes, my thought has been that if I support that I would try `if`-expressions with required `()` around the alternative expressions. For example: `y = if x<10 (x-1) else (x*42);` That's symmetric with `if`-statements with required `{}` around the alternative statements.

**Reply by MaxSagebaum** [reply]
The comprehensions in Python leave me always wondering, what is the type that is returned.  For a strongly types language I think, that it gets awkward and messy defining this right.

About:
`y = if x<10 (x-1) else (x*42);`

Considering that otherwise `{}` brackets are required, this would be a good option. But what when the if expression requires brackets? Then it would read:
`y = if (x<10) (x-1) else (x*42);`
which looks kind of weird. Here a `then` seems natural but I do not think it would be a good idea to introduce it:
`y = if (x<10) then (x-1) else (x*42);`

**Reply by JohelEGP** [reply]
> when the if expression requires brackets

Isn't that never?
Oh, there's the case of #632.

**Reply by MaxSagebaum** [reply]
You also might want to enclose it in brackets if the expressions becomes more complicated.

**Reply by ntrel** [reply]
With that syntax, when the test is a function call the true branch could look like it's calling the result of the function call:
```c++
y = if f(10) (x-1) else (x*42);
y = if g(10)(x~) (x-1) else (x*42);
```
I think using `=` is clearer:
```c++
y = if f(10) = x-1 else x*42;
y = if g(10)(x~) = x-1 else x*42;

y = if x<10 = x-1 else x*42; // 1
```
We already use `= expression` for inspect expression alternatives:
```c++
y = inspect x<10 -> bool { is true = x-1; is _ = x*42; } // same as 1 above
y = inspect x -> bool { if x<10 = x-1; is _ = x*42; } // same but with `if` alternative from P2392
```
(Note: The alternative on the first line is parsed: `is` *postfix-expression* `=`, so the `=` is not parsed as part of an expression. I have a PR for an `if`*logical-or-expression* `=` - alternative, see #753).

If an assignment should be allowed inside the condition, it has to be in parentheses:
```c++
x: int;
...
y = if (x = expr) < 10 = x-1 else x*42;
```

**Reply by msadeqhe** [reply]
Nice idea. To make it more similar to `inspect`, it would require `;` and another `=` inside parentheses:

```cpp
y = (if x < 10 = x - 1; else = x * 42);
```

**Reply by msadeqhe** [reply]
Also if we make comprehensions similar to `inspect`, it would be like this:

```cpp
list: vector<i32> = (1, 2, 3, 4, 5);
var2: = (for i in list = i * 2); // (2, 4, 6, 8, 10)
```

In general:

```cpp
var1: = (for i in list if i < 10 = i; else = i - 10);
```

**Reply by AbhinavK00** [reply]
I think the `;` before `else` is not needed because that gives the false impression that a line ends there but ofcourse it doesn't. And it does not affect readibility at all because keyword `else` is present.
Question: why the need for parentheses?
Edit: Also, just noticed that the `for` comprehension uses the keyword `in`, it probably should be rewritten to use `do` to make it consistent with for loops.

**Reply by msadeqhe** [reply]
Parentheses prevent `;` to end the outer statement, and `;` is for consistency with `inspect` syntax.

You're right about comprehension syntax. It should use `do` instead of `in`.

**Reply by AbhinavK00** [reply]
Why not just get rid of parentheses and semi-colon to save 3 characters?
Edit: I think python-style syntax is just better, it reads naturally and saves requirement for symbols like `=`, `;` etc.

Just look at this
```
var1: = i for list do i if i < 10 else i - 10;
```
Point out any problems you see with this, those can be discussed but I think it's already tried and tested and therefore the better choice.

**Reply by msadeqhe** [reply]
Yes. Python-style is more expressive unless consistency with Cpp2 syntax is a concern.

**Reply by msadeqhe** [reply]
Alternative syntax with `?`:

```cpp
a: = if x < 10 ? x else x - 10;
b: = for list do (i) i * 2;
```

Also `do (i) i * 2` is similar to terser lambda syntax but it's without `:`.

**Reply by AbhinavK00** [reply]
Using a full character for a small feature isn't a good idea I think.

**Reply by msadeqhe** [reply]
You're right, but `?` can be extended to other control structures too:

```cpp
a: = while x < 10 next i++ ? i * 2;
```

BTW `?` is already used in Cpp1. It's just another alternative syntax.

**Reply by msadeqhe** [reply]
Using `()` in *if-expressions* with direct and list initializations:

```cpp
v32: type == std::vector<i32>;

// With required `()`
var1: v32 = if x < 10 ((1, 2)) else ((1, 2, 3));

// With optional `()`
var2: v32 = if x < 10 (1, 2) else (1, 2, 3);
```

**Comment by jcanizales** [comment]
What would the type of `var2` be?

**Reply by msadeqhe** [reply]
Its return type would be `std::generator<i32>`. It's like writing a coroutine function with `co_yield`.

**Comment by msadeqhe** [comment]
- Discussion #836 is a generalized feature based on issue #824 and the idea of @ntrel to use `=` in control flows. I close this discussion in favor of block expressions. Thanks everyone.
